feat: separate jwt-set attributes from manual-set

edpaget opened the pull request:
Closes https://github.com/metabase/metabase/issues/21682

### Description

Saves attributes from a JWT into a separate jwt_attributes field that is merged with lower priority than manually set user attributes. It serialize the attributes back using the same mechanism as tenant attributes which allows us to provide a reasonable editing experience. This also updates the front-end attribute editor to handle the jwt attributes in the same way as tenant attributes.

This is adapted from the work that targeted the changes for tenants with all the tenanting removed. See: https://github.com/metabase/metabase/pull/61160

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

edpaget committed:
feat: separate jwt-set attributes from manual-set

iethree committed:
allow tenant attributes CRUD (#60347)

wip merge user and tenant attributes

WIP use new api

Split Mapping Editor

types fixed

so much unit testing

handle null cases

update tests

update a couple more e2e tests

less silly test

make things nicer

invalidate user cache on tenant update

edpaget committed:
fixup! typescript stuff

edpaget committed:
feat: restore structrue-attributes merging

Co-Authored-by: Dan Sutton <dan@metabase.com>

edpaget committed:
feat: use merged attributes in routing/sandbox/etc

Co-Authored-By: John Swanson <john.swanson@metabase.com>

edpaget committed:
fixup! move group manager test to ee

edpaget committed:
fixup! kondo

edpaget committed:
fixup! modal spec

edpaget committed:
fixup! cypress tests

edpaget committed:
fixup! migration timestamp

edpaget committed:
fixup! one more cherry pick oopsie

edpaget committed:
fixup! sandboxing cypress helper fixes

edpaget committed:
fixup! more cypress updates

edpaget committed:
fixup! remove eslint directive

johnswanson commented on enterprise/backend/src/metabase_enterprise/database_routing/common.clj:
```diff
@@ -22,7 +22,7 @@
   ([user db-or-id]
    (when-let [attr-name (user-attribute db-or-id)]
-     (let [database-name (get (:login_attributes user) attr-name)]
+     (let [database-name (get (api/current-user-attributes) attr-name)]
```
I don't think this will have any ill effects in practice because I've noticed we're never calling this function with a user other than `api/*current-user*`, but seems like a footgun to ignore the passed user here.

One option might be to just make sure all uses of this function call it with the 1-arity version (only the `db-or-id` without the user) and then remove the 2-arity version since it doesn't seem to be needed in practice. I think the function is only called in a few places so it shouldn't be that hard (I think I did this for the tenant attributes bits).

edpaget replied:
Yeah almost all the calls to this function were passing in @api/*current-user* it looked like.

johnswanson reviewed (changes requested):
One small suggested change but other than that LGTM.

edpaget committed:
fixup! remove 2-arity router-dbor-id->destination-db-id

All the places it was called with 2-arity were explicitly passing in
@api/current-user anyway

johnswanson reviewed (approved):
Looks good!
